R-package/src/lightgbm_R.cpp

```diff
@@ -1271,9 +1372,24 @@ static const R_CallMethodDef CallEntries[] = {
   {NULL, NULL, 0}
 };
 
-LIGHTGBM_C_EXPORT void R_init_lightgbm(DllInfo *dll);
+LIGHTGBM_C_EXPORT void R_init_lib_lightgbm(DllInfo *dll);
 
-void R_init_lightgbm(DllInfo *dll) {
+void R_init_lib_lightgbm(DllInfo *dll) {
```

Comment:
I don't think this and the similar change above it are necessary. I suspect those might be the cause of the MSVC failures you're seeing.

Can you please revert them? This function should be called `R_init_lightgbm`, for both the CRAN (`autotools`) package and the CMake-based builds.

As described in https://cran.r-project.org/doc/manuals/R-exts.html

> *It is good practice for DLLs to register their symbols (see [Registering native routines](https://cran.r-project.org/doc/manuals/R-exts.html#Registering-native-routines)), restrict visibility (see [Controlling visibility](https://cran.r-project.org/doc/manuals/R-exts.html#Controlling-visibility)) and not allow symbol search (see [Registering native routines](https://cran.r-project.org/doc/manuals/R-exts.html#Registering-native-routines)). It should be possible for a DLL to have only one visible symbol, R_init_pkgname, on suitable platforms[84](https://cran.r-project.org/doc/manuals/R-exts.html#FOOT84), which would completely avoid symbol conflicts.*

See also:

https://github.com/microsoft/LightGBM/blob/48e3629dc69ca6b28690933f600be2a00cc7f031/R-package/src/lightgbm-win.def#L3

and #4494

Reply:
It would fail in my setup if I rename the init function to not have `lib_`. Changing it in the altrep registration name doesn't seem to generate any failures though, so I've changed it there to see if it makes a difference.

Reply:
I'd be very surprised to learn that this function name needs to change as a part of this PR. It's already a part of the C API for `lib_lightgbm.dll` (from the perspective of the R package built with MSVC), and the changes in this PR are all otherwise not exposed as C API changes.

At a minimum, I suspect the CI failures you're seeing for MSVC are because you're changing this function name but not changing the `lightgbm-win.def` file I linked above.

Reply:
In the other cmake-based jobs that were tested here, leaving the name as `R_init_lightgbm` makes the job fail, while the same jobs succeed when changing it to `R_init_lib_lightgbm`.

The `-win.def` file that you mentioned AFAIK is not used in the msvc build from cmake.

Reply:
hmmmm ok. Thanks for looking into that! 

I still don't understand though what about these changes would require that function name to be changed to `R_init_lib_lightgbm` when it is working on `master` as `R_init_lightgbm`. Given that you're just adding new behavior *inside* that function.

Anyway, maybe this and the `.def` file are not the right place to look anyway.... looking more closely at the failing builds, I see that quite a few unit tests are passing before those jobs eventually fail.

e.g. on https://github.com/microsoft/LightGBM/actions/runs/7413984613/job/20174104011?pr=6213

```text
----- end of build and install logs -----
Running tests with testthat.R
basic: ........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
custom_objective: ..........
dataset: ................................S.............................................
learning_to_rank: ....................................................
lgb.Booster: ..........................................................................
Error: Process completed with exit code -1.
```

I have a Windows laptop with MSVC and R on it.... I can pull this branch and test. Maybe I'll at least be able to get some logs pointing to what is happening. Powershell provided on GitHub Actions runners (which we use to orchestrate those tests, see https://github.com/microsoft/LightGBM/blob/master/.ci/test_r_package_windows.ps1) is very picky about how it manages stderr and we've put in many changes to work around that in the future. For example:

https://github.com/microsoft/LightGBM/blob/48e3629dc69ca6b28690933f600be2a00cc7f031/.ci/test_r_package_windows.ps1#L16-L29

Maybe some useful error or warning is being written to stderr and we're not seeing it in CI logs as a result of those changes.

Thanks for all your efforts so far! I can take over the investigation for a bit. I think it's likely the issue is about the peculiarities of LightGBM's CI setup.

Reply:
In the non-msvc cmake ones, before the name change to `R_init_lib_lightgbm`, the tests would fail when they try to do something with the variables representing altrep classes introduced here (`R_altrep_class_t`), which would not have been initialized by the time the `.Call` uses them and thus have invalid pointers.

Reply:
So maybe a start could be to verify whether lines like this one are run before `.Call`:
```cpp
lgb_altrepped_char_vec = R_make_altraw_class("lgb_altrepped_char_vec", "lightgbm", dll);
```

Reply:
Iiiiinteresting, thanks for that!

Since you mention this being maybe related to the timing of when those statements run, I think it's important to point out a difference between the MSYS/MinGW CMake CI jobs and the MSVC ones.

The MSYS/MinGW ones run `R CMD check`.

https://github.com/microsoft/LightGBM/blob/48e3629dc69ca6b28690933f600be2a00cc7f031/.ci/test_r_package_windows.ps1#L133

https://github.com/microsoft/LightGBM/blob/48e3629dc69ca6b28690933f600be2a00cc7f031/.ci/test_r_package_windows.ps1#L179

The MSVC ones just run `Rscript --vanilla R-package/tests/testthat.R`

https://github.com/microsoft/LightGBM/blob/48e3629dc69ca6b28690933f600be2a00cc7f031/.ci/test_r_package_windows.ps1#L281-L288

That could result in different code paths being followed and a different order of things being initialized? I'm not sure. If that's the case, maybe we'll have to add a `.onLoad()` or `.onAttach()` to the package (though I really hope not).

I'll look into that.

Reply:
I'm going to work on removing this difference in the library name (`lib_lightgbm` vs. `lightgbm)` in the different types of R builds here. Work in progress at #6432. Hopefully it'll help with this PR.

Reply:
#6432 has been merged, so:

* all this `lib_lightgbm` vs. `lightgbm` stuff in the R package is gone
* the CMake-based builds now *should* do routine registration and invoke `R_init_lightgbm()` just like the CRAN-style build do

Let's see in CI if it helps.